xds/internal/resolver/cluster_specifier_plugin_test.go

```diff
@@ -20,349 +20,466 @@ package resolver
-	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed.Config) {
-		t.Errorf("ClientConn.UpdateState received different service config")
-		t.Error("got: ", cmp.Diff(nil, rState.ServiceConfig.Config))
-		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed.Config))
-	}
-
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes:  []*xdsresource.Route{{Prefix: newStringP(""), ClusterSpecifierPlugin: "cspA"}},
-			},
-		},
-		// Top level csp config here - the value of cspA should get directly
-		// placed as a child policy of xds cluster manager.
-		ClusterSpecifierPlugins: map[string]clusterspecifier.BalancerConfig{"cspA": []map[string]interface{}{{"csp_experimental": cspConfig{ArbitraryField: "changed"}}}},
-	}, nil)
-
-	gotState, err = tcc.stateCh.Receive(ctx)
-	if err != nil {
-		t.Fatalf("Error waiting for UpdateState to be called: %v", err)
+		t.Fatalf("Received error in service config: %v", rState.ServiceConfig.Err)
 	}
-	rState = gotState.(resolver.State)
-	if err := rState.ServiceConfig.Err; err != nil {
-		t.Fatalf("ClientConn.UpdateState received error in service config: %v", rState.ServiceConfig.Err)
-	}
-	wantJSON = `{"loadBalancingConfig":[{
-    "xds_cluster_manager_experimental":{
-      "children":{
-        "cluster_specifier_plugin:cspA":{
-          "childPolicy":[{"csp_experimental":{"arbitrary_field":"changed"}}]
-        }
-      }
-    }}]}`
-
-	wantSCParsed = internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(wantJSON)
+	wantSCParsed = internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(`
+{
+	"loadBalancingConfig": [
+		{
+		  "xds_cluster_manager_experimental": {
+			"children": {
+			  "cluster_specifier_plugin:cspA": {
+				"childPolicy": [
+				  {
+					"csp_experimental": {
+					  "arbitrary_field": "changed"
+					}
+				  }
+				]
+			  }
+			}
+		  }
+		}
+	  ]
+}`)
 	if !internal.EqualServiceConfigForTesting(rState.ServiceConfig.Config, wantSCParsed.Config) {
-		t.Errorf("ClientConn.UpdateState received different service config")
-		t.Error("got: ", cmp.Diff(nil, rState.ServiceConfig.Config))
-		t.Fatal("want: ", cmp.Diff(nil, wantSCParsed.Config))
+		t.Fatalf("Got service config:\n%s \nWant service config:\n%s", cmp.Diff(nil, rState.ServiceConfig.Config), cmp.Diff(nil, wantSCParsed.Config))
 	}
 }
 
 // TestXDSResolverDelayedOnCommittedCSP tests that cluster specifier plugins and
 // their corresponding configurations remain in service config if RPCs are in
 // flight.
 func (s) TestXDSResolverDelayedOnCommittedCSP(t *testing.T) {
-	xdsR, xdsC, tcc, cancel := testSetup(t, setupOpts{target: target})
-	defer xdsR.Close()
-	defer cancel()
-
+	// Env var GRPC_EXPERIMENTAL_XDS_RLS_LB controls whether the xDS client
+	// allows routes with cluster specifier plugin as their route action.
```

Comment:
Same comment as above. Although I guess you can leave them, since the env var talks about features enabled/disabled, not implementation specifics of how it goes about it.

Reply:
Replied to the other comment.